File: MATLAB/Rx/check_sig_field.m
function [valid] = check_sig_field(rate,res,length,parity,tail)

if res ~= 0 % check reserved bit is set to zero
    valid = 0;
elseif tail ~= zeros(1,6) % check tail is 6 zeros
    valid = 0;
elseif rem(sum([rate,length,parity]),2) ~= 0 % even parity check
    valid = 0;
else
    if sum(rate == [1 1 0 1]) == 4
        valid = 1;
    elseif sum(rate == [1 1 1 1]) == 4
        valid = 1;
    elseif sum(rate == [0 1 0 1]) == 4
        valid = 1;
    elseif sum(rate == [0 1 1 1]) == 4
        valid = 1;
    elseif sum(rate == [1 0 0 1]) == 4
        valid = 1;
    elseif sum(rate == [1 0 1 1]) == 4
        valid = 1;
    elseif sum(rate == [0 0 0 1]) == 4
        valid = 1;
    elseif sum(rate == [0 0 1 1]) == 4
        valid = 1;
    else
        valid = 0;
    end
end

File: PYTHON/RX/check_sig_field.py
import numpy as np


# The purpose of this function is to check that the SIG field values determined from sig_field_decoder() are valid
# such that they follow the correct format of the plcp described in the 802.11a - 1999 standard

def check_sig_field(rate, res, length, parity, tail):
    # convert each field to a numpy array:
    rate = np.array(rate)
    rate = rate.astype(int)
    length = np.array(length)
    length = length.astype(int)
    tail = np.array(tail)
    tail = tail.astype(int)

    null_tail = np.zeros((1, 6))

    valid = 0

    if res != 0:
        valid = 0
    elif tail is null_tail:
        valid = 0
    elif np.remainder((sum(rate) + sum(length) + parity), 2) != 0:
        valid = 0
    else:
        if np.array_equal(rate, np.array([1, 1, 0, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([1, 1, 1, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([0, 1, 0, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([0, 1, 1, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([1, 0, 0, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([1, 0, 1, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([0, 0, 0, 1])):
            valid = 1
        elif np.array_equal(rate, np.array([0, 0, 1, 1])):
            valid = 1
        else:
            valid = 0
    return valid
